RipCurrent/ripcurrent.h

```diff
@@ -550,20 +605,83 @@ class UserCodeInstantiator<LHSTypes<LHS_TYPES...>, RHSTypes<RHS_TYPES...>, USER_
-    virtual ~Scope() = default;
-
     explicit Scope(const current::LazilyInstantiated<
                        EventConsumerInitializer<LHSTypes<LHS_TYPES...>, RHSTypes<RHS_TYPES...>, USER_CLASS>,
                        std::shared_ptr<EmitDestination<LHSTypes<RHS_TYPES...>>>>& lazy_instance,
                    std::shared_ptr<EmitDestination<LHSTypes<RHS_TYPES...>>> next)
-        : next_(next), spawned_user_class_instance_(lazy_instance.InstantiateAsUniquePtrWithExtraParameter(next_)) {}
+        : next_(next),
+          spawned_user_class_instance_(lazy_instance.InstantiateAsUniquePtrWithExtraParameter(next_)),
+          waitable_counters_(0u, 0u),
+          single_threaded_processor_(waitable_counters_, *spawned_user_class_instance_),
+          mmq_(single_threaded_processor_) {}
+
+    virtual ~Scope() {
+      // TODO(dkorolev): Consider other termination strategies, not just the plain "process everything" one.
+      waitable_counters_.Wait([](const ThreadMessageCounters& p) { return p.ProcessedEverything(); });
+    }
+
+    void OnThreadUnsafeEmitted(movable_message_t&& x, std::chrono::microseconds t) override {
+      waitable_counters_.MutableUse([](ThreadMessageCounters& p) { p.ReportMessagePublished(); });
+      try {
+        mmq_.Publish(std::move(x), t);
+      } catch (const ss::InconsistentTimestampException& e) {
+        current::Singleton<RipCurrentMockableErrorHandler>().HandleError(e.What());
+        waitable_counters_.MutableUse([](ThreadMessageCounters& p) { p.ReportMessageNotQuitePublished(); });
+      }
+    }
+
+    void OnThreadUnsafeScheduled(movable_message_t&& x, std::chrono::microseconds t) override {
+      waitable_counters_.MutableUse([](ThreadMessageCounters& p) { p.ReportMessagePublished(); });
+      try {
+        mmq_.PublishIntoTheFuture(std::move(x), t);
+      } catch (const ss::InconsistentTimestampException& e) {
+        current::Singleton<RipCurrentMockableErrorHandler>().HandleError(e.What());
+        waitable_counters_.MutableUse([](ThreadMessageCounters& p) { p.ReportMessageNotQuitePublished(); });
+      }
+    }
 
-    void OnEmitted(CurrentSuper&& x) override { spawned_user_class_instance_->Accept(std::move(x)); }
+    virtual void OnThreadUnsafeHeadUpdated(std::chrono::microseconds t) override {
+      try {
+        mmq_.UpdateHead(t);
+      } catch (const ss::InconsistentTimestampException& e) {
+        current::Singleton<RipCurrentMockableErrorHandler>().HandleError(e.What());
+      }
+    }
 
    private:
+    class ThreadMessageCounters {
+     public:
+      ThreadMessageCounters(int published, int processed) : published_(published), processed_(processed) {}
+      void ReportMessagePublished() { ++published_; }
+      void ReportMessageNotQuitePublished() { --published_; }
+      void ReportMessageProcessed() { ++processed_; }
+      bool ProcessedEverything() const { return published_ == processed_; }
+
+     private:
+      int published_;
+      int processed_;
```

Comment:
Are these variables declared signed intentionally? You initialize them above with `0u`.